scipy/optimize/_differentialevolution.py

```diff
@@ -272,12 +284,25 @@ def differential_evolution(func, bounds, args=(), strategy='best1bin',
 
     Now repeat, but with parallelization.
 
-    >>> bounds = [(0,2), (0, 2), (0, 2), (0, 2), (0, 2)]
     >>> result = differential_evolution(rosen, bounds, updating='deferred',
     ...                                 workers=2)
     >>> result.x, result.fun
     (array([1., 1., 1., 1., 1.]), 1.9216496320061384e-19)
 
+    If the function is vectorized, then this might be an alternative approach
+    to parallelization. Vectorization may require the objective function to be
+    rewritten slightly.
+
+    >>> def rosen_vec(x):
+    ...     # accept an (M, len(x0)) array, returning a (M,) array
+    ...     v = 100 * (x[..., 1:] - x[..., :-1]**2.0)**2.0
```

Comment:
NBD but ellipses suggests to me that there could be more dimensions. You might use `:` to emphasize that there will only be one more.
On the other hand, have you considered swapping the axes? `rosen` is already vectorized in that way, and it may require fewer changes for users to implement because `x[i]` etc... still works. (On the other hand, they might be more likely to make mistakes for that reason.) Anyway, just something to think about.

Reply:
> On the other hand, have you considered swapping the axes?

I'm trying to get my head around this at the moment. Internally the population array is stored as `(M, len(x))`, where `M` is the number of population members. This makes sense for a few reasons:

- C memory layout is row-major, so values for a given parameter vector are stored contiguously.
- You can iterate over M:  `energies = [func(x) for x in M]`.

However, as you point out it makes vectorisation in a function more complicated. Compare the two functions:

```
def f1(x):
    # accept an (M, len(x0)) array, returning a (M,) array
    return np.sin(x[..., 0]) + np.cos(x[..., 1])

def f2(x):
    # accept an (len(x0), M) array, returning a (M,) array
    return np.sin(x[0]) + np.cos(x[1])
```

Conceptually it's much simpler to write the latter. What would the users find less confusing?

Reply:
My guess is that users will find the latter less confusing. I think it's OK for internal representations to be different from what the user thinks about.

Reply:
I've changed around the order so that we use the second type of function. Furthermore, I've changed the nomenclature to be consistent with NonlinearConstraint.

N = number of parameters
S = number of solution vectors being examined
M = number of constraint components